_posts/2023/2023-06-06-getting-started-with-vector-dbs-in-python.md

```diff
@@ -221,8 +221,163 @@ USERNAME=admin PASSWORD=admin ENDPOINT=https://localhost:9200 poetry run src/ope
 {% endhighlight %}
 
+### Vespa
+
+[Vespa](https://vespa.ai/) is a fully featured search engine and vector database. It supports vector search (ANN), lexical search, and search in structured data, all in the same query. Vespa is Apache 2.0 licensed, and can be run in a variety of ways, including Docker and as a managed [cloud service](https://cloud.vespa.ai/).
+
+Let's use their Docker container for this example.
+
+{% highlight bash %}
+docker info | grep "Total Memory" # make sure it's at least 4Gb
+docker run --detach --name vespa --hostname vespa-container \
+  --publish 8080:8080 --publish 19071:19071 \
+  vespaengine/vespa
+{% endhighlight %}
+
+This container listsens on port `8080` for search and and ingestion APIs and on `19071` for configuration APIs.
+
+Vespa encapsulates the concept of a schema/index in an application that needs to be defined and deployed, so it is not as straightforward as the previous example.
+
+To create a new application with a sample vector schema we need to create a `settings.xml` file with the overall application properties and a `schema.md` file with the definition of our schema. For this example, let's create the following directory structure
+
+```
+vector-app/
+├── schemas/
+│   └── vector.sd
+└── services.xml
+```
+
+`services.xml`:
+{% highlight xml %}
+<?xml version="1.0" encoding="utf-8" ?>
+<services version="1.0" xmlns:deploy="vespa" xmlns:preprocess="properties">
+    <container id="default" version="1.0">
+        <document-api/>
+        <search/>
+        <nodes>
+            <node hostalias="node1" />
+        </nodes>
+    </container>
+    <content id="vector" version="1.0">
+        <redundancy>2</redundancy>
+        <documents>
+            <document type="vector" mode="index" />
+        </documents>
+        <nodes>
+            <node hostalias="node1" distribution-key="0" />
+        </nodes>
+    </content>
+</services>
+{% endhighlight %}
+
+`vector.sd`:
+{% highlight xml %}
+schema vector {
+    document vector {
+        field id type string {
+            indexing: summary | attribute
+        }
+        field values type tensor<float>(x[3]) {
+            indexing: summary | attribute
+            attribute {
+                distance-metric: angular
+            }
+        }
+        struct metadatatype {
+            field genre type string {}
+        }
+        field metadata type metadatatype {
+            indexing: summary
+        }
+    }
+    rank-profile vector_similarity {
+        inputs {
+            query(vector_query_embedding) tensor<float>(x[3])
+        }
+        first-phase {
+            expression: closeness(field, values)
+        }
+    }
+
+{% endhighlight %}
+
+We can then deploy using the configuration API:
```

Comment:
Let's be consistent with the above. We use full sentences, so just say "Deploy using the configuration API."

Reply:
Updated.